frontend/src/utils/index.js

```diff
@@ -704,3 +704,93 @@ export function targetText (target) {
+    return semver.diff(version, kubernetesVersion) === 'patch' && semver.gt(version, kubernetesVersion) && !isPreview
+  })
+}
+
+export function k8sVersionUpdatePathAvailable (kubernetesVersion, cloudProfileName) {
+  const allVersions = store.getters.kubernetesVersions(cloudProfileName)
+  if (k8sVersionIsNotLatestPatch(kubernetesVersion, cloudProfileName)) {
+    return true
+  }
+  const versionMinorVersion = semver.minor(kubernetesVersion)
+  return !!find(allVersions, ({ version, isPreview }) => {
+    return semver.minor(version) === versionMinorVersion + 1 && !isPreview
+  })
+}
+
+export function selectedImageIsNotLatest (machineImage, machineImages) {
+  const { version: testImageVersion, vendorName: testVendor } = machineImage
+
+  return !!find(machineImages, ({ version, vendorName, isPreview }) => {
+    return testVendor === vendorName && semver.gt(version, testImageVersion) && !isPreview
+  })
+}
+
+export function k8sExpirationForShoot (shootK8sVersion, shootCloudProfileName, k8sAutoPatch) {
+  const allVersions = store.getters.kubernetesVersions(shootCloudProfileName)
+  const version = find(allVersions, { version: shootK8sVersion })
+  if (!version || !version.expirationDate) {
+    return undefined
+  }
+
+  const patchAvailable = k8sVersionIsNotLatestPatch(shootK8sVersion, shootCloudProfileName)
+  const updatePathAvailable = k8sVersionUpdatePathAvailable(shootK8sVersion, shootCloudProfileName)
+  const updateAvailable = !patchAvailable && updatePathAvailable
+
+  const isError = !updatePathAvailable
+  const isWarning = !isError && ((!k8sAutoPatch && patchAvailable) || updateAvailable)
+  const isInfo = !isError && !isWarning && k8sAutoPatch && patchAvailable
+
+  if (!isError && !isWarning && !isInfo) {
+    return undefined
+  }
+  return {
+    expirationDate: version.expirationDate,
+    isValidTerminationDate: isValidTerminationDate(version.expirationDate),
+    isError,
+    isWarning,
+    isInfo
+  }
+}
+
+export function expiredWorkerGroupsForShoot (shootWorkerGroups, shootCloudProfileName, imageAutoPatch) {
+  const expiredWorkerGroups = []
+  const allMachineImages = store.getters.machineImagesByCloudProfileName(shootCloudProfileName)
+  forEach(shootWorkerGroups, worker => {
+    const workerImage = get(worker, 'machine.image')
+    const workerImageDetails = find(allMachineImages, workerImage)
+    const updateAvailable = selectedImageIsNotLatest(workerImageDetails, allMachineImages)
+
+    const isError = !updateAvailable
+    const isWarning = !imageAutoPatch && updateAvailable
+    const isInfo = imageAutoPatch && updateAvailable
+    if (workerImageDetails.expirationDate &&
+      (isError || isWarning || isInfo)) {
+      expiredWorkerGroups.push({
+        ...workerImageDetails,
+        isValidTerminationDate: isValidTerminationDate(workerImageDetails.expirationDate),
+        workerName: worker.name,
+        isError,
+        isWarning,
+        isInfo
+      })
+    }
+  })
```

Comment:
instead of a forEach you could first use map to create the workerGroups and then filter for expiredWorkerGroups. WDYT?

Reply:
done